File: solutions/ruby/beer-song/1/beer_song.rb
# frozen_string_literal: true

class BeerSong
  def self.recite(from, count)
    bottles = ->(n) { n.zero? ? "no more bottles" : (n > 1 ? "#{n} bottles" : "#{n} bottle") }

    from.downto(from-count+1).map do |n|
      n.zero? ?
        "No more bottles of beer on the wall, no more bottles of beer.\n"+
        "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
      : "#{bottles.(n)} of beer on the wall, #{bottles.(n)} of beer.\n".capitalize +
        "Take #{ n > 1 ? "one" : "it" } down and pass it around, #{bottles.(n-1)} of beer on the wall.\n"
    end.join("\n")
  end
end

File: solutions/clojure/beer-song/2/src/beer_song.clj
(ns beer-song)

(require '(clojure [string :as s]))

(defn verse
  "Returns the nth verse of the song."
  [num]
  (let [s (if (== num 2) "" "s")]
    (case num
      0 (str "No more bottles of beer on the wall, no more bottles of beer.\n"
             "Go to the store and buy some more, 99 bottles of beer on the wall.\n")
      1 (str "1 bottle of beer on the wall, 1 bottle of beer.\n"
             "Take it down and pass it around, no more bottles of beer on the wall.\n")
      (format (str "%d bottles of beer on the wall, %d bottles of beer.\n"
                   "Take one down and pass it around, %d bottle%s of beer on the wall.\n")
              num num
              (dec num) s))))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end] (->>
                (range start (dec end) -1)
                (map verse)
                (s/join "\n"))))
